lib/charms/opensearch/v0/opensearch_distro.py

```diff
@@ -305,13 +309,79 @@ def write_file(self, path: str, data: str, override: bool = True):
         with open(path, mode="w") as f:
             f.write(data)
 
+    @property
+    def _plugin(self):
+        """Returns the executable path. Should be overloaded, e.g. by the snap command."""
+        return f"{self.paths.home}/bin/opensearch-plugin"
+
+    def add_plugin_without_restart(self, plugin: str, batch: bool = False) -> bool:
```

Comment:
The plugins related methods should only be as part of the `PluginManager` and not have any say in the restart logic

```suggestion
    def add_plugin(self, plugin: str, batch: bool = False) -> bool:
```

Reply:
* Move the keystore and plugin install/remove/list to the PluginManager: remove from OpenSearchDistribution
* In case we need, we can extend the PluginManager to the k8s example, in the future
* The restart logic will be executed by the charm, as folllows, for example backup plugin:

There will be two files:
src/opensearch_backup.py
```
class OpenSearchCharmBackup(...):

     def __init__(self, charm):
          self.plugin_manager = self.charm.plugin_manager
          self.framework.observe(self.charm.on.s3_credentials_changed, self.s3_credentials_changed)

    def s3_credentials_changed(self, event):
         ...
         if self.plugin_manager.configure("repository-s3", ...):
              # needs a restart
```

In the main charm:
```

class OpenSearchCharm(BaseCharm):

    def __init__(self, ...):
        self.plugin_manager = OpenSearchPluginManager(...)
        self.framework.observe(self.charm.on.config_changed, self.on_config_changed)


    def on_config_changed(self, event):
        ....
        if self.plugin_manager.setUp(self.config): # the plugin manager figures out which to save
            # we need a restart
```